specs/core/0_beacon-chain.md

```diff
@@ -1629,59 +1629,107 @@ Below are the processing steps that happen at every `block`.
-* Verify that `bls_verify(pubkey=proposer.pubkey, message_hash=signed_root(proposer_slashing.proposal_2, "signature"), signature=proposer_slashing.proposal_2.signature, domain=get_domain(state.fork, slot_to_epoch(proposer_slashing.proposal_2.slot), DOMAIN_PROPOSAL))`.
-* Run `slash_validator(state, proposer_slashing.proposer_index)`.
+For each `proposer_slashing` in `block.body.proposer_slashings`, run the following function:
+
+```python
+def process_proposer_slashing(proposer_slashing: ProposerSlashing,
+                              state: BeaconState):
+    proposer = state.validator_registry[proposer_slashing.proposer_index]
+    # Verify that the slot is the same
+    assert proposer_slashing.proposal_1.slot == proposer_slashing.proposal_2.slot
+    # Verify that the shard is the same (or that both proposals are beacon chain proposals)
+    assert proposer_slashing.proposal_1.shard == proposer_slashing.proposal_2.shard
+    # But the roots are different!
+    assert proposer_slashing.proposal_1.block_root != proposer_slashing.proposal_2.block_root
+    # Proposer is not yet slashed
+    assert proposer.slashed is False
+    # Signatures are valid
+    for proposal in (proposer_slashing.proposal_1, proposer_slashing.proposal_2):
+        assert bls_verify(
+            pubkey=proposer.pubkey,
+            message_hash=signed_root(proposal, "signature"),           
+            signature=proposal.signature,
+            domain=get_domain(state.fork, slot_to_epoch(proposal.slot), DOMAIN_PROPOSAL)
+        )
+    slash_validator(state, proposer_slashing.proposer_index)
+```
 
 ##### Attester slashings
 
 Verify that `len(block.body.attester_slashings) <= MAX_ATTESTER_SLASHINGS`.
 
-For each `attester_slashing` in `block.body.attester_slashings`:
+For each `attester_slashing` in `block.body.attester_slashings`, run the following function:
 
-* Let `slashable_attestation_1 = attester_slashing.slashable_attestation_1`.
-* Let `slashable_attestation_2 = attester_slashing.slashable_attestation_2`.
-* Verify that `slashable_attestation_1.data != slashable_attestation_2.data`.
-* Verify that `is_double_vote(slashable_attestation_1.data, slashable_attestation_2.data)` or `is_surround_vote(slashable_attestation_1.data, slashable_attestation_2.data)`.
-* Verify that `verify_slashable_attestation(state, slashable_attestation_1)`.
-* Verify that `verify_slashable_attestation(state, slashable_attestation_2)`.
-* Let `slashable_indices = [index for index in slashable_attestation_1.validator_indices if index in slashable_attestation_2.validator_indices and state.validator_registry[index].slashed is False]`.
-* Verify that `len(slashable_indices) >= 1`.
-* Run `slash_validator(state, index)` for each `index` in `slashable_indices`.
+```python
+def process_attester_slashing(attester_slashing: ProposerSlashing,
+                              state: BeaconState):
```

Comment:
Prefer to re-order the parameters: `state` and then `attester_slashing` as most helper functions + fix type hinting.
```python
def process_attester_slashing(state: BeaconState,
                              attester_slashing: ProposerSlashing) -> None:
```